firebase_functions/functions/src/migration.ts

```diff
@@ -0,0 +1,150 @@
+import { Client, ClientBuilder } from '@iota/client';
+import * as admin from 'firebase-admin';
+import * as crypto from 'crypto';
+
+export const migration = async (from: number, to: number) => {
+  await deleteUnusedDevices();
+  await deleteUnusedUsers(from, to);
+  await updateDevices(from, to);
+  await updateUser(from, to);
+};
+
+const deleteUnusedDevices = async () => {
+  const deviceRef = admin.firestore().collection('devices');
+  await deviceRef.get().then(async (querySnapshot) => {
+    const docs = querySnapshot.docs;
+    console.log("Docs length: ", docs.length)
+    for (let docIndex = 50; docIndex < 174; docIndex++ ) {
+      const device = docs[docIndex];
+      console.log("Device: ", device.id)
+      const data = await admin.firestore().collection('devices').doc(device.id).collection('data').limit(1).get();
+      if (data.size === 0) {
+        console.log("------> Deleting device: ", device.id)
+        await admin.firestore().collection('devices').doc(device.id).delete();
+        await admin.firestore().collection('deviceList').doc(device.id).delete();
+      }
+    }
+  });
+};
+
+const deleteUnusedUsers = async (from: number, to: number) => {
+    const userRef = admin.firestore().collection('users');
+    await userRef.get().then(async (querySnapshot) => {
+      const docs = querySnapshot.docs;
+      console.log("Docs length: ", docs.length)
+      for (let docIndex = from; docIndex < to; docIndex++ ) {
+        const device = docs[docIndex];
+        if (!device) {
+            console.log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~", docIndex)
+            continue
+        }
+        console.log("User: ", device?.id)
+        const data = await admin.firestore().collection('users').doc(device.id).collection('purchases').limit(1).get();
+        if (data.size === 0) {
+          console.log("------> Deleting user: ", device.id)
+          await admin.firestore().collection('users').doc(device.id).delete();
+        }
+      }
+    });
+  };
+
+const updateDevices = async (from: number, to: number) => {
+  const deviceListRef = admin.firestore().collection('deviceList');
+  const deviceRef = admin.firestore().collection('devices');
+
+  await deviceListRef.get().then(async (querySnapshot) => {
+    const docs = querySnapshot.docs;
+    for (let docIndex = from; docIndex < to; docIndex++) {
```

Comment:
Since you are invoking a Promise-like construct (e.g. `await`) inside a for-loop, I would suggest to replace all such synchronous for-loops with it's async version
```
for await (const item of items) {
    await function(item);
};

```

Reply:
Didn't even know about for await loops. The solution is not as easy as copy and paste because I'm limiting the loop with the `from` and `to` parameters. Since it is just a migration script, that probably won't get used in the future I would not change the loop if it is okay for you.